### PR TITLE
Pre parsl

### PR DESCRIPTION
**Pull request by OliverEvans96**
These are changes I made before working on Parsl integration. I thought it would be good to break this up into two PRs. Hopefully I didn't break anything here.

### CODE REVIEW

File: kale/workflow_objects.py [hunk 1]
```diff
@@ -366,10 +396,36 @@ def draw_dag(self, layout=None):
...
         """Return DAG containing only steps which are to be run. (Not yet implemented.)"""
-        return self.dag
+        subdag = networkx.DiGraph()
+
+        if self._bqgraph.selected is None or len(self._bqgraph.selected) == 0:
+            return self.dag
+        else:
+            self.check_selection_for_jumps()
+
+            for node in self.dag.nodes():
```
Comment (mlhenderson) [inline kale/workflow_objects.py]: this section in gen_subdag should have comments describing the intent of the code for future reference

File: kale/workflow_objects.py [hunk 2]
```diff
@@ -366,10 +396,36 @@ def draw_dag(self, layout=None):
...
+                node_index = node.index[self]
+                if node_index in self._bqgraph.selected:
+                    subdag.add_node(node)
+                    parent_list = list(self.dag.predecessors(node))
+                    if len(parent_list) > 0:
```
Comment (mlhenderson) [inline kale/workflow_objects.py]: this if is not needed, the for loop will not execute if parent_list has 0 elements

File: kale/workflow_objects.py [hunk 3]
```diff
@@ -366,10 +396,36 @@ def draw_dag(self, layout=None):
...
 
     def gen_subdag(self):
         """Return DAG containing only steps which are to be run. (Not yet implemented.)"""
-        return self.dag
+        subdag = networkx.DiGraph()
+
+        if self._bqgraph.selected is None or len(self._bqgraph.selected) == 0:
+            return self.dag
+        else:
+            self.check_selection_for_jumps()
```
Comment (mlhenderson) [inline kale/workflow_objects.py]: should have a comment about what the behavior should be if gaps are detected

File: kale/workflow_objects.py [hunk 4]
```diff
@@ -366,10 +396,36 @@ def draw_dag(self, layout=None):
...
+
+            for node in self.dag.nodes():
+                node_index = node.index[self]
+                if node_index in self._bqgraph.selected:
```
Comment (mlhenderson) [inline kale/workflow_objects.py]: when would this be false?

File: kale/workflow_widgets.py [hunk 5]
```diff
@@ -571,21 +570,19 @@ def _call_read_log(self, caller=None):
...
-                    print("Workflow submitted.")
-                else:
-                    print("Workflow already running.")
+                self.futures.append(self._thread_pool.submit(self.run_workflow))
+                print("Workflow submitted.")
             except AttributeError as e:
                 # Attribute error if self.future is None
                 # which means workflow has not been submitted.
-                self.future = self._thread_pool.submit(self.run_workflow)
+                self.future = [self._thread_pool.submit(self.run_workflow)]
```
Comment (mlhenderson) [inline kale/workflow_widgets.py]: self.futures vs self.future? looks like this could be a typo

File: kale/workflow_objects.py (hunk 1 above)
Comment (mlhenderson) [reply kale/workflow_objects.py]: Wondering if you can iterate over self._bqgraph.selected instead, then check for ancestors and add edges?

File: kale/workflow_objects.py (hunk 4 above)
Comment (OliverEvans96) [reply kale/workflow_objects.py]: If the node is not selected

File: kale/workflow_widgets.py (hunk 5 above)
Comment (OliverEvans96) [reply kale/workflow_widgets.py]: Good catch!

File: kale/workflow_objects.py (hunk 1 above)
Comment (OliverEvans96) [reply kale/workflow_objects.py]: Good idea - I changed this

**Comment by OliverEvans96** [comment]
I agree that I was too quick to put everything in the source tree. I think some example .py files are useful as in [ipyvolume examples](https://github.com/maartenbreddels/ipyvolume/blob/master/ipyvolume/examples.py), but notebooks and data for examples should be outside of the source tree. Would it be appropriate to deal with this as a separate PR after the Parsl stuff is merged to avoid excess rebasing?

Also, I agree that the environment config is an important and separate topic, as is logging.

**Pull request merged**